python/rapidsmpf/rapidsmpf/integrations/single.py

```diff
@@ -0,0 +1,495 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+"""Shuffler integration for single-worker pylibcudf execution."""
+
+from __future__ import annotations
+
+import threading
+from typing import TYPE_CHECKING, Any, cast
+
+import rmm.mr
+from rmm.pylibrmm.stream import DEFAULT_STREAM
+
+from rapidsmpf.buffer.buffer import MemoryType
+from rapidsmpf.buffer.resource import BufferResource, LimitAvailableMemory
+from rapidsmpf.communicator.single import new_communicator
+from rapidsmpf.config import Optional, OptionalBytes, Options
+from rapidsmpf.integrations.core import ShufflerIntegration, WorkerContext
+from rapidsmpf.progress_thread import ProgressThread
+from rapidsmpf.rmm_resource_adaptor import RmmResourceAdaptor
+from rapidsmpf.shuffler import Shuffler
+
+if TYPE_CHECKING:
+    from collections.abc import Callable, Sequence
+
+    from rapidsmpf.integrations.core import DataFrameT
+
+
+# Set of available shuffle IDs
+_shuffle_id_vacancy: set[int] = set(range(Shuffler.max_concurrent_shuffles))
+_shuffle_id_vacancy_lock: threading.Lock = threading.Lock()
+
+
+# Local single-worker context
+class _SingleWorker:
+    """Mutable single-worker utility class."""
+
+    context: WorkerContext | None = None
+
+
+_single_rapidsmpf_worker: _SingleWorker = _SingleWorker()
+
+
+def get_single_worker_context() -> WorkerContext:
+    """
+    Retrieve the single `WorkerContext`.
+
+    If the worker context does not already exist on the worker, it
+    will be created and attached to `_single_rapidsmpf_worker_context`.
+
+    Returns
+    -------
+    The existing or newly initialized worker context.
+    """
+    with WorkerContext.lock:
```

Comment:
Is this lock necessary for correctness, or is it just a consequence of trying to reuse some code? Put differently, do we actually have multiple threads active here?

Reply:
It should be possible for multiple threads to call this function at the same time. I definitely saw issues in an earlier iteration of this branch without the lock. With that said, all tests seem to pass without the lock right now, so something may have changed.

Reply:
Thinking through this again: I think we do need a lock here if we're hoping to support multi-threaded applications that might independently launch shuffles. We somehow need to protect access to this shared, global resource.

I can't remember, but I think my original question was motivated by whether this was needed by a single shuffle in isolation (which I hope it doesn't, but can't say for sure).